File: backend/app/services/four_table/h3_account_scope.py
"""H3 投资性房地产的语义科目定位规格（单一真源）。

**修掉的真实缺陷（2026-08-01）**

`_h3_investment_property._H3_ACCOUNT_PREFIXES` 原写::

    {"1503": 投资性房地产原值, "1504": 累计折旧}

而 `account_chart` 实证：``1503 = 可供出售金融资产``（旧准则，G6 域）、
``1504 = 债权投资``（G4 域）。投资性房地产真实科目族是::

    1521 投资性房地产
    1525 投资性房地产累计折旧
    1526 投资性房地产累计摊销      ← 土地使用权走摊销，原实现完全没有这一槽
    1527 投资性房地产减值准备      ← 同样缺失

即 H3 取数**取的是另外两个循环的科目**，且缺两个槽。

**为什么必须走语义定位而不是把码改成 1521/1525/1526/1527**

`account_mapping` 只读实证（10 个项目）—— 同一原始码在不同项目映射到**不同标准码**::

    1525 投资性房地产累计折旧 → 1521（1 个项目，**并入母科目**）
    1525 投资性房地产累计折旧 → 1525（4 个项目，独立）
    1527 投资性房地产减值准备 → 1521（1 个项目）
    1527 投资性房地产减值准备 → 1527（3 个项目）

那个把累计折旧并入 `1521` 的项目，若按标准码 `1525` 查 `trial_balance` 会**取空**；
反之按 `1521` 查会把原值与累计折旧混在一起。故只能按**科目名称**在该项目自己的科目表里定位，
再各自经 `account_mapping` 换算 —— 这正是
`semantic_account_resolver` 存在的理由。

**报表行**（`report_config` 实证，四准则公式**不同**，不可只看一条）::

    BS-027 投资性房地产
      listed_consolidated : TB('1521','期末余额')
      listed_standalone   : TB('1521') - TB('1525') - TB('1526')
      soe_consolidated    : TB('1521','期末余额')
      soe_standalone      : TB('1521') - TB('1525')
    IMP-010 九、投资性房地产减值准备
      soe_standalone      : TB('1527','期末余额')

`report_config` 在本循环**恰好是对的**，故它作为「提示 + 冲突检测」会与语义结果一致；
一旦某项目自定义了报表行公式，`SemanticAccountResult.conflicts` 会把差异暴露到溯源面板。

🔴 **`direction` 不可用于备抵判定**：实证 `1525`/`1526`/`1527` 在
`account_chart` 里 `direction` 全是 ``debit``（不是 credit）→ 只能靠科目名关键字。
`is_provision` 因此在槽里**显式声明**，不由方向推断。

spec: .kiro/specs/g-cycle-extraction-mapping-and-disclosure-alignment/
      Requirements 2.1~2.3 / Property 4
"""
from __future__ import annotations

from .semantic_account_resolver import SemanticAccountSlot, SemanticAccountSpec

#: 原值槽的否决词 —— `投资性房地产累计折旧` **包含** `投资性房地产`，
#: 无否决词时原值会把三个备抵槽一并吃掉（虚增原值）。
_GROSS_EXCLUDES = (
    "累计折旧",
    "累计摊销",
    "减值准备",
    "减值损失",
    "公允价值变动",
    "处置",
)

#: 输出槽键 → 前端 `tb_values` 键前缀（保持既有契约：`ip_*` / `dep_*` 原样不动）
H3_SLOT_KEY_PREFIX = {
    "gross": "ip",
    "accum_dep": "dep",
    "accum_amort": "amort",
    "impairment": "impair",
}

H3_ACCOUNT_SPEC = SemanticAccountSpec(
    row_code="BS-027",
    slots=(
        SemanticAccountSlot(
            key="gross",
            names=("投资性房地产",),
            exclude_names=_GROSS_EXCLUDES,
            fallback_standard_codes=("1521",),
            label="投资性房地产原值",
        ),
        # 🔴🔴 **不得**把裸名 `累计折旧` / `累计摊销`列为兜底名（2026-08-04 真实库实证）
        #
        # `match_slot_in_chart` 按 `names` 声明顺序取第一个**精确命中**的名字。
        # DB 实证：裸名 `累计折旧` 在全库唯一对应 `1602`（**固定资产**累计折旧，
        # standard 侧 10 个项目 / client 侧 7 个项目都是这个裸名），裸名 `累计摊销`
        # 唯一对应 `1702`（无形资产累计摊销）；而 `1525 投资性房地产累计折旧` /
        # `1526 投资性房地产累计摊销` 只在 4~5 个项目的科目表里存在。
        # → 科目表缺 1525/1526 时，原实现的第二个名字会精确命中 1602/1702，
        #   把**固定资产的累计折旧 + 无形资产的累计摊销**从投资性房地产原值里扣掉，
        #   实测三个项目的投资性房地产账面金额变成负数
        #   （`4f6dbc36` −21,601,944.08 / `df5b8403` −11,322,704.22 /
        #    `f064f5e4` −21,864,702.78）。`exclude_names` 拦不住 —— 裸名里没有
        #   「固定资产」「无形资产」这些关键词。
        #
        # 删掉裸名后：科目表有 1525/1526 → 按专名精确命中（行为不变）；
        # 没有 → 槽 `found=False` → 如实按「本项目无此科目」处理、**不扣减**
        #（宁缺勿造），账面金额回归原值口径。
        SemanticAccountSlot(
            key="accum_dep",
            names=("投资性房地产累计折旧",),
            exclude_names=("固定资产", "使用权资产", "生产性生物资产"),
            fallback_standard_codes=("1525",),
            label="累计折旧",
            is_provision=True,
        ),
        SemanticAccountSlot(
            key="accum_amort",
            names=("投资性房地产累计摊销",),
            exclude_names=("无形资产", "使用权资产", "长期待摊费用"),
            fallback_standard_codes=("1526",),
            label="累计摊销",
            is_provision=True,
        ),
        SemanticAccountSlot(
            key="impairment",
            names=("投资性房地产减值准备",),
            exclude_names=("减值损失",),
            fallback_standard_codes=("1527",),
            label="减值准备",
            is_provision=True,
        ),
    ),
    # 旧准则同族科目：客户仍在用时提示人工映射，**不自动归槽**
    # （旧准则「投资性房地产」科目号本身就是 1521，故此处主要防御自建科目表的变体命名）
    legacy_standard_names=(),
)

__all__ = ["H3_ACCOUNT_SPEC", "H3_SLOT_KEY_PREFIX"]

File: audit-platform/frontend/src/components/workpaper/composables/h3AccountScope.ts
/**
 * h3AccountScope — H3 投资性房地产科目定位的**前端单一真源**（per-cycle 薄壳）。
 *
 * **要修的缺陷（2026-08-06 实证，数字级 P0）**
 *
 * 后端 `four_table/h3_account_scope.py` 已于 2026-08-01 把 H3 的科目族从
 * `1503`/`1504` 纠正为 `1521`/`1525`/`1526`/`1527`（`1503` 实为**可供出售金融资产**
 * 属 G6 域、`1504` 实为**债权投资**属 G4 域），而**前端仍全线写死旧码**：
 *
 * | 落点 | 旧实现 | 后果 |
 * |------|--------|------|
 * | `H3TabAdjudicationCost.writebackTrialBalance()` | `account_code: '1503'` / `'1504'` | **往 G6/G4 的科目行写投资性房地产审定数** —— 与 H8/H9 那次「往 1901 写审定数污染 K2」同款 |
 * | `H3TabAdjudicationFair.writebackTrialBalance()` | `account_code: '1503'` | 同上 |
 * | `useAdjudicationBringIn({subjectPrefix})` ×4 | `'1503'`/`'1504'`/`'1505'` | 从集中登记拉**别的循环**的调整分录 |
 * | `GtVoucherSamplingEngine account-code` | `'1503'` | 抽的是可供出售金融资产的凭证 |
 * | `eventBus substantive:adjudicated` | `accountCode: '1503'` | 下游相关性判定挂错科目 |
 * | AI `related_data.accountCode` | `'1503'` | 给模型的上下文是错科目 |
 *
 * **口径优先级（与平台既有 `k2AccountScope` / `g6AccountScope` 一致）**
 *
 * 运行态一律取 render 下发的 `html_data.tb_source_codes`（后端逐项目按**科目名称**
 * 在该项目自己的 `account_chart` 里定位的结果）；本文件的常量**只作兜底与展示**。
 *
 * 🔴 为什么不能只把字面量从 1503 换成 1521：`account_mapping` 实证同一原始码在不同
 * 项目映射到**不同标准码**（`1525 投资性房地产累计折旧 → 1521` 1 个项目并入母科目 /
 * `→ 1525` 4 个项目独立），按标准码硬查会在部分项目取空。故必须走 render 下发值。
 *
 * spec: .kiro/specs/h-cycle-extraction-formula-and-disclosure-completion/
 *       Requirements 5.1~5.4（溯源面板）与 Task 19（前端旧码清零）
 */
import {
  createCycleAccountScope,
  type CycleAccountScopeSpec,
} from './shared/cycleAccountScope'

/** 后端槽键（须与 `h3_account_scope.py` 的 `H3_ACCOUNT_SPEC.slots` 逐字一致） */
export const H3_SLOT_GROSS = 'gross'
export const H3_SLOT_ACCUM_DEP = 'accum_dep'
export const H3_SLOT_ACCUM_AMORT = 'accum_amort'
export const H3_SLOT_IMPAIRMENT = 'impairment'

/**
 * 除主槽外要在溯源面板明细里展示的槽。
 *
 * 🔴 H3 是**四槽** spec 且**没有** `provision` 槽 ⇒
 * `SemanticAccountResult.as_dict()` 的扁平 `provision` 恒为空，
 * 不显式声明这三个槽键，累计折旧/累计摊销/减值准备的来源科目在面板里看不见。
 */
export const H3_EXTRA_SLOT_KEYS = [
  H3_SLOT_ACCUM_DEP,
  H3_SLOT_ACCUM_AMORT,
  H3_SLOT_IMPAIRMENT,
] as const

/** 报表行（`report_config` 实证：BS-027 投资性房地产 / IMP-010 减值准备） */
export const H3_REPORT_ROW_CODE = 'BS-027'
export const H3_IMPAIRMENT_REPORT_ROW_CODE = 'IMP-010'

export const H3_ACCOUNT_SCOPE_SPEC: CycleAccountScopeSpec = {
  cycle: 'H3',
  reportRowCode: H3_REPORT_ROW_CODE,
  slots: [
    { key: H3_SLOT_GROSS, label: '投资性房地产原值', fallback: '1521' },
    { key: H3_SLOT_ACCUM_DEP, label: '累计折旧', fallback: '1525', isProvision: true },
    { key: H3_SLOT_ACCUM_AMORT, label: '累计摊销', fallback: '1526', isProvision: true },
    { key: H3_SLOT_IMPAIRMENT, label: '减值准备', fallback: '1527', isProvision: true },
  ],
}

export const h3AccountScope = createCycleAccountScope(H3_ACCOUNT_SCOPE_SPEC)

/**
 * 兜底展示码（**仅**在 render 未下发时用于文案，勿作为查询口径）。
 *
 * 🔴 这些是「本循环正确的科目族」，与被替换掉的 `1503`/`1504`/`1505` 完全不同 ——
 * 后者分属 G6（可供出售金融资产）/ G4（债权投资）/（债权投资减值准备）三个别的循环。
 */
export const H3_FALLBACK_CODES = {
  gross: '1521',
  accumDep: '1525',
  accumAmort: '1526',
  impairment: '1527',
} as const

export type H3SlotKey =
  | typeof H3_SLOT_GROSS
  | typeof H3_SLOT_ACCUM_DEP
  | typeof H3_SLOT_ACCUM_AMORT
  | typeof H3_SLOT_IMPAIRMENT

/**
 * `html_data.tb_values` 的键前缀 —— **必须与后端 `H3_SLOT_KEY_PREFIX` 逐字一致**。
 *
 * 🔴 改造前 `useH3FormData` 读的是 `inv_prop_1503_unadjusted` / `dep_1504_unadjusted`，
 * 而后端 `build_h3_tb_values` 产出的是 `ip_*` / `dep_*` / `amort_*` / `impair_*`
 * ⇒ 那四个键是 **dead render key**，seed 分支永不命中、必然落到按 1503/1504 的
 * 错误 HTTP 兜底查询。守卫 `h3AccountScopeContract.spec.ts` 读后端 py 源码交叉锁死。
 */
export const H3_TB_PREFIX = {
  gross: 'ip',
  accumDep: 'dep',
  accumAmort: 'amort',
  impairment: 'impair',
} as const
